# 1.3 Formulating Abstractions with Higher-Order Procedures

## 1.3.1 Procedures as Arguments

## Example 9: Summation

Consider the following three procedures.  
The first computes the $sum$ of the integers from $a$ through $b$:

The second computes the $sum$ of the $cubes$ of the integers in the given range:

The third computes the $sum$ of a sequence of $terms$ in the series  

$$\frac{1}{1 \centerdot 3} + \frac{1}{5 \centerdot 7} + \frac{1}{9 \centerdot 11} + \ldots$$

which converges to $\pi$/8 (very slowly):

Indeed, mathematicians long ago identified the abstraction of $summation$ of a series and invented "sigma notation", for example,to express this concept.

$$\sum_{n=a}^bf(n) = f(a) + \ldots+ f(b)$$

With the aid of the procedures sum as follows,we can express the $mathematical$ $concepts$ of $summation$:

In [1]:
cat 1.3/Example_9/sum_by_recursion.scm

(define (sum term a next b)
  (if (> a b) 0
      (+ (term a)
         (sum term (next a) next b))))


$\bullet$ We can redefine the process $sum$-$integers$ to compute the $sum$ of the integers from $a$ through $b$:

In [3]:
cat 1.3/Example_9/sum_integers.scm

(load "1.3/Example_9/sum_by_recursion.scm")

(define (identity x) x)
(define (inc n) (+ n 1))

(define (sum-integers a b)
  (sum identity a inc b))


### Running Instance:

$\bullet$ Redefine the process $sum$-$cubes$ to compute the sum of the cubes of the integers from a through b:

In [4]:
cat 1.3/Example_9/sum_cubes.scm

(load "1.3/Example_9/sum_by_recursion.scm")

(define (cube x) (* x x x))
(define (inc n) (+ n 1))

(define (sum-cubes a b)
  (sum cube a inc b))


### Running Instance:

$\bullet$ We can also redefine $pi$-$sum$ in the same way:

In [5]:
cat 1.3/Example_9/pi_sum.scm

(load "1.3/Example_9/sum_by_recursion.scm")

(define (pi-sum a b)
  (define (pi-term x)
    (/ 1.0 (* x (+ x 2))))
  (define (pi-next x) (+ x 4))
  (sum pi-term a pi-next b))


### Running Instance:

Once we have $sum$, we can use it as a building block in formulating further concepts. For instance, the $definite$ $integral$ of a function $f$ between the limits $a$ and $b$ can be approximated numerically using the formula

$$\int_a^bf=\left[f\left(a + \frac{dx}{2}\right) + f\left(a + dx + \frac{dx}{2}\right) + f\left(a + 2dx + \frac{dx}{2}\right) + \centerdot\centerdot\centerdot\right]dx$$

for small values of $dx$. We can express this directly as a procedure:

In [6]:
cat 1.3/Example_9/definite_integral.scm

(load "1.3/Example_9/sum_by_recursion.scm")

(define (integral f a b dx)
  (define (add-dx x) (+ x dx))
  (* (sum f (+ a (/ dx 2.0)) add-dx b) dx))


For example,we can let $f$ is $cube$ function:

### Running Instance:

#### The exact value of the integral of cube between 0 and 1 is 1/4.

## Exercise 1.29:
Simpson’s Rule is a more accurate method of numerical integration than the method illustrated above. Using Simpson’s Rule, the integral of a function $f$ between $a$ and $b$ is approximated as

$$\frac{h}{3}(y_0 + 4y_1 + 2y_2 + 4y_3 + 2y_4 + \cdots + 2y_{n-2} + 4y_{n-1} + y_n)$$

where $h = (b -a)/n$,for some even integer n, and $y_k = f(a + kh)$.(Increasing n increases the accuracy of the approximation.)Define a procedure that takes as arguments $f$ , $a$, $b$, and $n$ and returns the value of the integral, computed using Simpson’s Rule.Use your procedure to integrate cube between 0 and 1 (with $n = 100$ and $n = 1000$), and compare the results to those of the integral procedure shown above.

### Answer:

In [7]:
cat 1.3/Exercise_1.29/simpson.scm

(load "1.3/Example_9/sum_by_recursion.scm")

(define (simpson f a b n)
  (define h (/ (- b a) n))
  (define (y k) (f (+ a (* k h))))
  (define (factor k)
    (cond ((or (= k 0) (= k n)) 1)
          ((odd? k) 4)
          (else 2)))
  (define (term k) (* (factor k) (y k)))
  (define (next k) (+ k 1))
  (if (not (even? n))
      (error "n can't be odd number.")
      (* (/ h 3) (sum term (exact->inexact 0) next n))))


### Running Instance:

## Exercise 1.30: 
The $sum$ procedure above generates a $linear$ $recursion$. The procedure can be rewritten so that the sum is performed $iteratively$. Show how to do this by filling in the missing expressions in the following definition:

### Answer:

In [8]:
cat 1.3/Exercise_1.30/sum_by_iteration.scm

(define (sum term a next b)
  (define (iter a result)
    (if (> a b)
        result
        (iter (next a)
              (+ (term a) result))))
  (iter a 0))
		


## Exercise 1.31:
a. The $sum$ procedure is only the simplest of a vast number of similar abstractions that can be captured as higher-order procedures.Write an analogous procedure called $product$ that returns the product of the values of a function at points over a given range. Show how to define $factorial$ in terms of $product$. Also use product to compute approximations to $\pi$ using the formula

$$\frac{\pi}{4} = \frac{2\centerdot4\centerdot4\centerdot6\centerdot6\centerdot8\cdots}{3\centerdot3\centerdot5\centerdot5\centerdot7\centerdot7\cdots}$$

b. If your product procedure generates a recursive process,write one that generates an iterative process. If it generates an iterative process, write one that generates a recursive process.

### Answer:

$\bullet$ The $product$ procedure as follows

In [9]:
cat 1.3/Exercise_1.31/product_by_recursion.scm

(define (product term a next b)
  (if (> a b) 1
      (* (term a)
         (product term (next a) next b))))


$\bullet$ Redefine $factorial$ procedure in terms of $product$ procedure

In [11]:
cat 1.3/Exercise_1.31/factorial_by_product_recursion.scm

(load "1.3/Exercise_1.31/product_by_recursion.scm")

(define (identity x) x)
(define (inc n) (+ n 1))
(define (factorial b)
  (product identity 1 inc b))


### Running Instance:

$\bullet$ Use $product$ procedure to compute approximations to $\pi$

In [13]:
cat 1.3/Exercise_1.31/pi_by_product_recursion.scm

(load "1.3/Exercise_1.31/product_by_recursion.scm")

(define (product-pi a b)
  (define (pi-next n) (+ n 2))
  (define (pi-term a)
    ( * 1.0 (/ (* a (pi-next a)) (square (+ a 1)))))
  (product pi-term a pi-next b))


### Running Instance:

Since the product procedure above generates a recursive process;we should write one that generates an $iterative$ process

In [14]:
cat 1.3/Exercise_1.31/product_by_iteration.scm

(define (product term a next b)
  (define (iter a result)
    (if (> a b) result
        (iter (next a) (* (term a) result))))
  (iter a 1))


## Exercise 1.32:
a. Show that $sum$ and $product$ (Exercise 1.31) are both special cases of a still more general notion called $accumulate$ that combines a collection of terms, using some general accumulation function:

$accumulate$ takes as arguments the same term and range specifications as $sum$ and $product$, together with a $combiner$ procedure (of two arguments) that specifies how the current term is to be combined with the accumulation of the preceding terms and a null-value that specifies what base value to use when the terms run out. Write $accumulate$ and show how $sum$ and $product$ can both be defined as simple calls to $accumulate$.

b. If your $accumulate$ procedure generates a recursive process, write one that generates an iterative process.If it generates an iterative process, write one that generates a recursive process.

### Answer:

$\bullet$ The $accumulate$ recursive procedure:

In [15]:
cat 1.3/Exercise_1.32/accumulate_by_recursion.scm

(define (accumulate combiner null-value term a next b)
  (if (> a b) null-value
      (combiner (term a)
                (accumulate combiner
                            null-value
                            term
                            (next a)
                            next
                            b))))


$\bullet$ Redefine the $sum$ procedure by way of the $accumulate$ recursive procedure:

In [16]:
cat 1.3/Exercise_1.32/sum_by_accumulate_recursion.scm

(load "1.3/Exercise_1.32/accumulate_by_recursion.scm")

(define (sum term a next b)
  (accumulate + 0 term a next b))


$\bullet$ Redefine the $product$ procedure by way of the $accumulate$ recursive procedure:

In [17]:
cat 1.3/Exercise_1.32/product_by_accumulate_recursion.scm

(load "1.3/Exercise_1.32/accumulate_by_recursion.scm")

(define (product term a next b)
  (accumulate * 1 term a next b))


### Test 1: Summation

In [1]:
cat 1.3/Exercise_1.32/sum_integers_by_accumulate_recursion.scm

(load "1.3/Exercise_1.32/sum_by_accumulate_recursion.scm")

(define (identity x) x)
(define (inc n) (+ n 1))
(define (sum-integers a b)
  (sum identity a inc b))


### Running Instance:

### Test 2: Product

In [3]:
cat 1.3/Exercise_1.32/pi_product_by_accumulate_recursion.scm

(load "1.3/Exercise_1.32/product_by_accumulate_recursion.scm")

(define (pi-next n) (+ n 2))
(define (pi-term a)
  (* 1.0 (/ (* a (pi-next a)) (square (+ a 1)))))
(define (product-pi a b)
  (product pi-term a pi-next b))


### Running Instance:

The $accumulate$ $iterative$ procedure:

In [4]:
cat 1.3/Exercise_1.32/accumulate_by_iteration.scm

(define (accumulate combiner null-value term a next b)
  (define (iter a result)
    (if (> a b) result
        (iter (next a)
              (combiner (term a) result))))
  (iter a null-value))


## Exercise 1.33:
You can obtain an even more general version of $accumulate$ (Exercise 1.32) by introducing the notion of a $filter$ on the terms to be combined. That is, combine only those terms derived from values in the range that satisfy a specified condition. The resulting $filtered$-$accumulate$ abstraction takes the same arguments as $accumulate$, together with an additional predicate of one argument that specifies the $filter$. Write filtered accumulate as a procedure.Show how to express the following using $filtered$-$accumulate$:

### Answer:

#### A.The filtered-accumulate recursive procedure:

In [5]:
cat 1.3/Exercise_1.33/filtered_accumulate_by_recursion.scm

(define (filtered-accumulate combine null-value term a next b valid)
  (if (> a b) null-value
      (if (valid a)
         (combine (term a)
                  (filtered-accumulate combine
                                       null-value
                                       term
                                       (next a)
                                       next
                                       b
                                       valid))
        (filtered-accumulate combine
       	                     null-value
       	                     term
       	                     (next a)
       	                     next
       	                     b
       	                     valid))))


#### a.
the $sum$ of the squares of the prime numbers in the interval a to b (assuming that you have a $prime$ predicate already written)

In [7]:
cat 1.3/Exercise_1.33/sum_square_of_prime_by_filtered_accumulate_by_recursion.scm

(load "1.3/Exercise_1.33/filtered_accumulate_by_recursion.scm")
(load "1.2/Example_8/prime_by_search_divisor.scm")

(define (inc n) (+ n 1))
(define (sum-prime a b)
  (filtered-accumulate + 0 square a inc b prime?))


### Running Instance:

#### b. 
the $product$ of all the positive integers less than $n$ that are relatively prime to $n$  
(i.e., all positive integers $i \lt n$ such that $GCD(i,n) = 1$).

In [8]:
cat 1.3/Exercise_1.33/product_of_coprimes.scm

(load "1.3/Exercise_1.33/filtered_accumulate_by_recursion.scm")
(load "1.2/Example_7/gcd_by_Euclid_Algorithm.scm")

(define (identity x) x)
(define (coprime a b) (and (< a b) (= 1 (gcd a b))))
(define (inc n) (+ n 1))
(define (product-of-coprimes n)
  (filtered-accumulate * 1 identity 1 inc n (lambda (x) (coprime x n))))


### Running Instance:

#### B.The filtered-accumulate iterative procedure:

In [9]:
cat 1.3/Exercise_1.33/filtered_accumulate_by_iteration.scm

(define (filtered-accumulate combine null-value term a next b valid)
  (define (iter i result)
    (cond ((> i b) result)
          ((valid i)
           (iter (next i) (combine (term i) result)))
          (else (iter (next i) result))))
  (iter a null-value))


#### a.
the sum of the cubes of the prime numbers in the interval a to b:

In [10]:
cat 1.3/Exercise_1.33/sum_cube_of_prime_by_filtered_accumulate_by_iteration.scm

(load "1.3/Exercise_1.33/filtered_accumulate_by_iteration.scm")
(load "1.2/Example_8/prime_by_search_divisor.scm")

(define (inc n) (+ n 1))
(define (cube x) (* x x x))
(define (sum-prime a b)
  (filtered-accumulate + 0 cube a inc b prime?))


### Running Instance:

#### b.
the product of all the positive even integers less than n to n:

In [12]:
cat 1.3/Exercise_1.33/sum_even_integers_by_filtered_accumulate_by_iteration.scm

(load "1.3/Exercise_1.33/filtered_accumulate_by_iteration.scm")

(define (identity x) x)
(define (inc n) (+ n 1))
(define (sum-even-integers n)
  (filtered-accumulate + 0 identity 1 inc n even?))


### Runing Instance:

## 1.3.2 Constructing Procedures Using lambda

## Example 10: Summation by lambda

In using $sum$ as in Section 1.3.1, it seems terribly awkward to have to define trivial procedures such as $pi$-$term$ and $pi$-$next$ just so we can use them as arguments to our higher-order procedure. Rather than define $pi$-$next$ and $pi$-$term$, it would be more convenient to have a way to directly specify “the procedure that returns its input incremented by 4” and “the procedure that returns the reciprocal of its input times its input plus 2.” We can do this by introducing the special form $lambda$, which creates procedures. Using $lambda$ we can describe what we want as

(lambda (x) (+ x 4))

and

Then our $pi$-$sum$ procedure can be expressed without defining any $auxiliary$ procedures as

In [13]:
cat 1.3/Example_10/pi_sum_by_lambda_recursion.scm

(load "1.3/Example_9/sum_by_recursion.scm")

(define (pi-sum a b)
  (sum (lambda (x) (/ 1.0 (* x (+ x 2))))
       a
       (lambda (x) (+ x 4))
       b))


### Running Instance:

Again using $lambda$, we can write the $integral$ procedure without having to define the auxiliary procedure $add$-$dx$:

In [14]:
cat 1.3/Example_10/definite_integral_by_lambda.scm

(load "1.3/Example_9/sum_by_recursion.scm")

(define (integral f a b dx)
  (* (sum f
          (+ a (/ dx 2.0))
          (lambda (x) (+ x dx))
          b)
     dx))


### Running Instance:

In general, $lambda$ is used to create procedures in the same way as $define$, except that no name is specified for the procedure:

The resulting procedure is just as much a procedure as one that is created using $define$.   
The only difference is that it has not been associated with any name in the environment. In fact,

is equivalent to

Like any expression that has a procedure as its value, a $lambda$ expression can be used as the operator in a combination such as

## Using let to create local variables

Another use of $lambda$ is in creating $local$ $variables$.We often need $local$ $variables$ in our procedures other than those that have been bound as $formal$ $parameters$.For example, suppose we wish to compute the function

$$f(x,y) = x(1 + xy)^2 + y(1 - y) + (1 + xy)(1 -y),$$

which we could also express as

$$a = 1 + xy,$$
$$b = 1 - y,$$
$$f(x,y) = xa^2 + yb + ab.$$

In writing a procedure to compute $f$ , we would like to include as $local$ $variables$ not only $x$ and $y$ but also the names of intermediate quantities like $a$ and $b$. One way to accomplish this is to use an auxiliary procedure to bind the $local$ $variables$:

In [15]:
cat 1.3/Example_11/bind_local_variables_by_auxiliary.scm

(define (f x y)
  (define (f-helper a b)
    (+ (* x (square a))
       (* y b)
       (* a b)))
  (f-helper (+ 1 (* x y)) (- 1 y)))


Of course, we could use a $lambda$ expression to specify an anonymous procedure for binding our $local$ $variables$. The body of $f$ then becomes a single call to that procedure:

In [16]:
cat 1.3/Example_11/bind_local_variables_by_lambda.scm

(define (f x y)
  ((lambda (a b)
    (+ (* x (square a))
    (* y b)
    (* a b)))
  (+ 1 (* x y))
  (- 1 y)))


This construct is so useful that there is a special form called $let$ to make its use more convenient.  
Using $let$, the $f$ procedure could be written as

In [17]:
cat 1.3/Example_11/bind_local_variables_by_let.scm

(define (f x y)
  (let ((a (+ 1 (* x y)))
        (b (- 1 y)))
    (+ (* x (square a))
    (* y b)
    (* a b))))


The general form of a $let$ expression is

The first part of the $let$ expression is a list of $name$-$expression$ pairs.  
When the $let$ is evaluated, each $name$ is associated with the value of the corresponding $expression$.   
The body of the $let$ is evaluated with these names bound as $local$ $variables$.   
The way this happens is that the $let$ expression is interpreted as an alternate syntax for

No new mechanism is required in the interpreter in order to provide $local$ $variables$.   
A $let$ expression is simply $syntactic$ $sugar$ for the underlying $lambda$ application

We can see from this equivalence that the scope of a variable specified by a $let$ expression is the body of the $let$.This implies that:

$\bullet$ $let$ allows one to bind variables as locally as possible to where they are to be used. For example, if the value of $x$ is 5, the value of the expression

is 38. Here, the $x$ in the body of the $let$ is 3, so the value of the let expression is 33.   
On the other hand, the $x$ that is the second argument to the outermost $+$ is still 5.

$\bullet$ The variables’values are computed outside the $let$. This matters when the expressions that provide the values for the $local$ $variables$ depend upon variables having the same names as the local variables themselves. For example, if the value of $x$ is 2, the expression

will have the value 12 because, inside the body of the $let$, $x$ will be 3 and $y$ will be 4 (which is the outer $x$ plus 2).

Sometimes we can use $internal$ $definitions$ to get the same effect as with $let$.   
For example, we could have defined the procedure $f$ above as

In [18]:
cat 1.3/Example_11/bind_local_variables_by_internal.scm

(define (f x y)
  (define a (+ 1 (* x y)))
  (define b (- 1 y))
  (+ (* x (square a))
     (* y b)
     (* a b)))


We prefer,however,to use $let$ in situations like this and to use $internal$ $define$ only for $internal$ $procedures$.

## Exercise 1.34: 
Suppose we define the procedure

Then we have

What happens if we (perversely) ask the interpreter to evaluate the combination $(f \;\, f)$? Explain.

### 解答：

根据上述的一般过程定义

等价于

则我们可以将题中的定义

改写成$lambda$形式

然后展开$(f\;\,f)$的执行过程

执行到 (2 2) 这一步时，解释器试图以2作为函数进行调用，但是2并不是一个过程  
所以执行出错并打印信息： The object 2 is not applicable 

## 1.3.3 Procedures as General Methods

## Example 12: Finding roots of equations by the half-interval method

The $half$-$interval$ $method$ is a simple but powerful technique for finding roots of an equation $f(x) = 0$, where $f$ is a continuous function. The idea is that, if we are given points $a$ and $b$ such that $f(a) < 0 < f(b)$,then $f$ must have at least one zero between $a$ and $b$. To locate a zero,let $x$ be the average of $a$ and $b$, and compute $f(x)$. If $f(x) > 0$, then $f$ must have a zero between $a$ and $x$. If $f(x) < 0$, then $f$ must have a zero between $x$ and $b$. Continuing in this way, we can identify smaller and smaller intervals on which $f$ must have a zero. When we reach a point where the interval is small enough, the process stops. Since the interval of uncertainty is reduced by half at each step of the process, the number of steps required grows as $\Theta(log(L/T))$, where $L$ is the length of the original interval and $T$ is the error tolerance (that is, the size of the interval we will consider “small enough”). Here is a procedure that implements this strategy:

$search$ is awkward to use directly, because we can accidentally give it points at which $f$’s values do not have the required sign, in which case we get a wrong answer. Instead we will use search via the following procedure, which checks to see which of the endpoints has a negative function value and which has a positive value, and calls the search procedure accordingly. If the function has the same sign on the two given points, the half-interval method cannot be used, in which case the procedure signals an error.

We can group the processes above:

In [20]:
cat 1.3/Example_12/half_interval_method.scm

(define (average a b) (/ (+ a b) 2.0))
(define (close-enough x y) (< (abs (- x y)) 0.001))

(define (search f neg-point pos-point)
  (let ((midpoint (average neg-point pos-point)))
    (if (close-enough neg-point pos-point) midpoint
      (let ((test-value (f midpoint)))
        (cond ((positive? test-value) (search f neg-point midpoint))
              ((negative? test-value) (search f midpoint pos-point))
              (else midpoint))))))

(define (half-interval-method f a b)
  (let ((a-value (f a)) (b-value (f b)))
    (cond ((and (negative? a-value) (positive? b-value)) (search f a b))
          ((and (negative? b-value) (positive? a-value)) (search f b a))
          (else (error "Values are not of opposite sign" a b)))))


### Running Instance:

## Example 13: Finding fixed points of functions

A number $x$ is called a $fixed$ $point$ of a function $f$ if $x$ satisfies the equation $f(x) = x$. For some functions $f$ we can locate a $fixed$ $point$ by beginning with an initial $guess$ and applying $f$ repeatedly,

$$f(x),\;\;\;f(f(x)),\;\;\;f(f(f(x))),\;\;\;\ldots,$$

until the value does not change very much. Using this idea, we can devise a procedure $fixed$-$point$ that takes as inputs a function and an initial guess and produces an approximation to a $fixed$ $point$ of the function. We apply the function repeatedly until we find two successive values whose difference is less than some prescribed tolerance:

In [21]:
cat 1.3/Example_13/fixed_point_of_function.scm

(define tolerance 0.00001)
(define (fixed-point f first-guess)
  (define (close-enough v1 v2)
    (< (abs (- v1 v2)) tolerance))
  (define (try guess)
    (let ((next (f guess)))
      (if (close-enough guess next)
          next
          (try next))))
  (try first-guess))


### Running Instance:

For example, we can use this method to approximate the $fixed$ $point$ of the $cosine$ function, starting with 1 as an initial approximation:

Similarly, we can find a solution to the equation $y\,=\,siny\;+\;cosy$:

The $fixed$-$point$ process is reminiscent of the process we used for finding $square$ $roots$ in Section 1.1.7. Both are based on the idea of repeatedly improving a guess until the result satisfies some criterion. In fact, we can readily formulate the $square$-$root$ computation as a $fixed$-$point$ search. Computing the $square$ $root$ of some number $x$ requires finding a $y$ such that $y^2 = x$. Putting this equation into the equivalent form $y = x/y$,we recognize that we are looking for a fixed point of the function $y\mapsto x=y$, and we can therefore try to compute $square$ $roots$ as

Unfortunately, this $fixed$-$point$ $search$ does not converge. Consider an initial guess $y_1$. The next guess is $y_2 = x/y_1$ and the next guess is $y_3 = x/y_2 = x/(x/y_1) = y_1$. This results in an infinite loop in which the two guesses $y_1$ and $y_2$ repeat over and over, oscillating about the answer.

One way to control such oscillations is to prevent the guesses from changing so much. Since the answer is always between our guess $y$ and $x/y$, we can make a new guess that is not as far from $y$ as $x/y$ by averaging $y$ with $x/y$, so that the next guess after $y$ is $\frac{1}{2}(y + x/y)$ instead of $x/y$. The process of making such a sequence of guesses is simply the process of looking for a fixed point of $y\mapsto\frac{1}{2}(y + x/y)$:

(Note that $y = \frac{1}{2}(y + x/y)$ is a simple transformation of the equation $y = x/y$; to derive it, add y to both sides of the equation and divide by 2.)

With this modification, the $square$-$root$ procedure works. In fact, if we unravel the definitions, we can see that the sequence of approximations to the square root generated here is precisely the same as the one generated by our original $square$-$root$ procedure of Section 1.1.7. This approach of averaging successive approximations to a solution, a technique that we call $average$ $damping$, often aids the convergence of $fixed$-$point$ searches.

## Exercise 1.35: 
Show that the $golden$ $ratio$ $\varphi$ (Section 1.2.2) is a $fixed$ $point$ of the transformation $x \mapsto 1 + 1/x$, and use this fact to compute $\varphi$ by means of the $fixed$-$point$ procedure.

### Answer:

In [23]:
cat 1.3/Exercise_1.35/golden_ratio.scm

(load "1.3/Example_13/fixed_point_of_function.scm")

(define golden-ratio
  (fixed-point (lambda (x) (+ 1 (/ 1 x))) 1.0))


### Running Instance:

1 ]=> (load "1.3/Exercise_1.34/golden_ratio.scm")

;Loading "1.3/Exercise_1.34/golden_ratio.scm"...
;  Loading "1.3/Example_13/fixed_point_of_function.scm"... done
;... done
;Value: golden-ratio

1 ]=> golden-ratio

;Value: 1.6180327868852458

## Exercise 1.36: 
Modify $fixed$-$point$ so that it prints the sequence of approximations it generates, using the newline and display primitives shown in Exercise 1.22. Then find a solution to $x^x = 1000$ by finding a fixed point of $x \mapsto log(1000)/log(x)$. (Use Scheme’s primitive $log$ procedure, which computes natural logarithms.) Compare the number of steps this takes with and without $average$ $damping$. (Note that you cannot start $fixed$-$point$ with a guess of 1, as this would cause division by $log(1) = 0$.)

### 解答：

$\bullet$ 修改Example_13中的fixed-point-of-function过程，使其能够打印生成的近似值序列

In [27]:
cat 1.3/Exercise_1.36/fixed_point_of_function.scm

;;Add 2-9 lines
(define (display-info guess step)
  (display "Step: ")
  (display step)
  (display " ")

  (display "Guess: ")
  (display guess)
  (newline))

(define tolerance 0.00001)
(define (fixed-point f first-guess)
  (define (close-enough v1 v2)
    (< (abs (- v1 v2)) tolerance))
  (define (try guess step)
    (display-info guess step)
    (let ((next (f guess)))
      (if (close-enough guess next)
          (begin
            (display-info next (+ 1 step))
            next)
          (try next (+ 1 step)))))
  (try first-guess 1))


$\bullet$ 我们也需要平均阻尼过程

In [19]:
cat 1.3/Exercise_1.36/average_damp.scm

(define (average a b) (/ (+ a b) 2.0))
(define (average-damp f) (lambda (x) (average x (f x))))


$\bullet$ 将 $x^x = 1000$形式化为$x \mapsto log(1000)/log(x)$

In [20]:
cat 1.3/Exercise_1.36/formula.scm

(define formula (lambda (x) (/ (log 1000) (log x))))


### Running Instance:

#### I. 不带average-damp过程的计算过程：

#### II. 带average-damp过程的计算过程：

对比上述两个计算过程，不使用average-damp过程的计算共35步；而使用average-damp过程的计算共10步；  
由此可见，使用average-damp过程能使得计算过程快速收敛。

## Exercise 1.37:

a. An infinite $continued$ $fraction$ is an expression of the form

$$f = \cfrac{N_1}{D_1 + \cfrac{N_2}{D_2 + \cfrac{N_3}{D_3 + \cdots}}}$$

As an example, one can show that the infinite $continued$ $fraction$ expansion with the $N_i$ and the $D_i$ all equal to 1 produces $1/\varphi$, where $\varphi$ is the $golden$ $ratio$ (described in Section 1.2.2). One way to approximate an infinite continued fraction is to truncate the expansion after a given number of terms. Such a truncation—a so-called $k$-$term$ $finite$ $continued$ $fraction$—has the form

$$\cfrac{N_1}{D_1 + \cfrac{N_2}{\ddots + \cfrac{N_k}{D_k}}}.$$

Suppose that $n$ and $d$ are procedures of one argument (the term index $i$) that return the $N_i$ and $D_i$ of the terms of the continued fraction. Define a procedure $cont$-$frac$ such that evaluating ($cont$-$frac$ $n$ $d$ $k$) computes the value of the $k$-$term$ finite continued fraction.Check your procedure by approximating $1/\varphi$ using

for successive values of $k$. How large must you make $k$ in order to get an approximation that is accurate to 4 decimal places?

b. If your $cont$-$frac$ procedure generates a recursive process, write one that generates an iterative process. If it generates an iterative process, write one that generates a recursive process.

### 解答：

#### I.递归版本的连分式过程

因为连分式本质上就是一个除法计算序列，所以题目给出$\textbf{k}$项连分式

$$\cfrac{N_1}{D_1 + \cfrac{N_2}{\ddots + \cfrac{N_k}{D_k}}}.$$

可以转换成以下等价的除法计算序列

$$N_1/(D_1+(N_2/(D_2+\cdots+(N_k/D_k))))$$

而这个除法序列又可以用一个递归表达式来表示

$$
cf(1) \\
N_1/(D_1+cf(2)) \\
N_1/(D_1 + (N_2/(D_2 + cf(3)))) \\
N_1/(D_1 + (N_2/(D_2 + (N_3/(D_3 + cf(4)))))) \\
\cdots \\
N_1/(D_1 + (N_2/(D_2 + (N_3/(D_3 + \cdots + (N_k/D_k))))))
$$

其中函数$cf(i)$表示连分式的第$\textbf{i}$个项

根据上述递归表达式，我们可以给出连分式过程的递归过程$cont$-$frac$-$by$-$recursion$

In [29]:
cat 1.3/Exercise_1.37/continued_fraction_by_recursion.scm

(define (cont-frac N D k)
  (define (cf i)
    (if (= k i)
        (/ (N k) (D k))
        (/ (N i) (+ (D i) (cf (+ i 1))))))
  (cf 1))


#### II.迭代版本的连分式过程

一个$\textbf{k}$项连分式等价于除法计算序列

$$N_1/(D_1+(N_2/(D_2+\cdots+(N_k/D_k))))$$

如果要迭代地计算这个除法计算序列，我们必须倒转公式中各个除法项计算的先后顺序，先计算高次项，然后才是低次项

也即是说，我们必须先计算

$$cf(k)=N_k/D_k$$

然后计算

$$cf(k-1)=N_{k-1}/(D_{k-1} + cf(k))$$

再然后计算

$$cf(k-2)=N_{k-2}/(D_{k-2} + cf(k-1))$$

直到到达

$$N_1/(D_1 + cf(2))$$

这时我们就得出了整个连分式的解，而且整个计算过程是迭代地进行的

根据上述迭代过程，我们可以给出连分式过程的迭代过程$cont$-$frac$-$by$-$iteration$

In [30]:
cat 1.3/Exercise_1.37/continued_fraction_by_iteration.scm

(define (cont-frac N D k)
  (define (iter i result)
    (if (= i 0) result
        (iter (- i 1) (/ (N i) (+ (D i) result)))))
  (iter (- k 1) (/ (N k) (D k))))


#### III.连分式定义的黄金分割率函数

In [31]:
cat 1.3/Exercise_1.37/continued_fraction_by_recursion.scm

(define (cont-frac N D k)
  (define (cf i)
    (if (= k i)
        (/ (N k) (D k))
        (/ (N i) (+ (D i) (cf (+ i 1))))))
  (cf 1))


### Running Instance:

In [32]:
cat 1.3/Exercise_1.37/golden_ratio_iteration.scm

(load "1.3/Exercise_1.37/continued_fraction_by_iteration.scm")
(define (golden-ratio k)
  (+ 1
     (cont-frac (lambda (i) 1.0)
                (lambda (i) 1.0)
                k)))


### Running Instance:

## Exercise 1.38: 
In 1737, the Swiss mathematician Leonhard Euler published a memoir $De$ $Fractionibus$ $Continuis$, which included a continued fraction expansion for $e$ $-$ $2$, where $e$ is the base of the natural logarithms. In this fraction, the $N_i$ are all $1$, and the $D_i$ are successively $1, 2, 1, 1, 4, 1, 1, 6, 1, 1, 8, \ldots$. Write a program that uses your $cont$-$frac$ procedure from Exercise 1.37 to approximate $e$, based on Euler’s expansion.

### 解答：

根据题中提示，我们列出$\textbf{i}$和$\textbf{$N_i$}$以及$\textbf{$D_i$}$的对应序列

\begin{aligned}
i\;\;\;\;& 1\;\;\;2\;\;3\;\;4\;\;5\;\;6\;\;7\;\;8\;\;9\;\;10\;\;11\;\;\dots \\
D_i\;\;\;\;& 1\;\;2\;\;1\;\;1\;\;4\;\;1\;\;1\;\;6\;\;1\;\;1\;\;8\;\;\dots
\end{aligned}

观察以上序列，可以发现$\textbf{$D_i$}$的规律：

$\bullet$ 当$\textbf{(i+1)}$取模$\textbf{3}$等于$\textbf{0}$时，$\textbf{$D_i$}$等于$\textbf{(i + 1)/3 * 2}$

$\bullet$ 其他情况下，$\textbf{$D_i$}$均等于$\textbf{1}$

In [33]:
cat 1.3/Exercise_1.38/e.scm

(load "1.3/Exercise_1.37/continued_fraction_by_iteration.scm")
(define (e k)
  (define (N i) 1)
  (define (D i)
    (if (= 0 (remainder (+ i 1) 3))
        (* 2 (/ (+ i 1) 3))
        1))
  (+ 2.0 (cont-frac N D k)))


### Running Instance:

## Exercise 1.39: 
A continued fraction representation of the $tangent$ function was published in 1770 by the German mathematician J.H. Lambert:

$$\tan x = \cfrac{x}{1 - \cfrac{x^2}{3 - \cfrac{x^2}{5 - \ldots}}}$$

where $x$ is in radians. Define a procedure $(tan-cf\;x\;k)$ that computes an approximation to the $tangent$ function based on Lambert’s formula. $k$ specifies the number of terms to compute, as in Exercise 1.37.

### 解答：

根据题意我们可以给出$\textbf{N}$与$\textbf{D}$的变化规律  
其中$\textbf{$N_i$}$只有$\textbf{i}$为$\textbf{1}$时返回$x$，其他情况下返回$-x^2$    
$\textbf{$D_i$}$返回$\textbf{i}$所指定位置的奇数,即当$\textbf{i}$为$\textbf{i}$时，返回$\textbf{1}$，$\textbf{i}$为$\textbf{2}$时，返回$\textbf{3}$，以此类推

In [1]:
cat 1.3/Exercise_1.39/tan_cf.scm

(load "1.3/Exercise_1.37/continued_fraction_by_iteration.scm")

(define (tan-cf x k)
  (define (N i) (if (= i 1) x (- (square x))))
  (define (D i) (- (* i 2) 1))
  (exact->inexact (cont-frac N D k)))


使用MIT Scheme内置的$tan$函数和$tan$-$cf$进行对比测试验证$tan$-$cf$的正确性

### Running Instance:

## 1.3.4 Procedures as Returned Values

## Example 14: Square Root -- Fixed Point

The above examples demonstrate how the ability to pass procedures as arguments significantly enhances the expressive power of our programming language. We can achieve even more expressive power by creating procedures whose returned values are themselves procedures.

We can illustrate this idea by looking again at the fixed-point example described at the end of Section 1.3.3. We formulated a new version of the square-root procedure as a fixed-point search, starting with the observation that $\sqrt{x}$ is a fixed-point of the function $y \mapsto x/y$.Then we used average damping to make the approximations converge. Average damping is a useful general technique in itself. Namely, given a function $f$, we consider the function whose value at $x$ is equal to the average of $x$ and $f(x)$.

We can express the idea of average damping by means of the following procedure:

In [8]:
cat 1.3/Example_14/average_damp.scm

(define (average a b) (/ (+ a b) 2))
(define (average-damp f) (lambda (x) (average x (f x))))


$average$-$damp$ is a procedure that takes as its argument a procedure $f$ and returns as its value a procedure (produced by the lambda) that,when applied to a number x, produces the average of $x$ and $(f\;x)$. For example, applying $average$-$damp$ to the square procedure produces a procedure whose value at some number $x$ is the average of 
$x$ and $x^2$.Applying this resulting procedure to 10 returns the average of 10 and 100, or 55

Using $average$-$damp$, we can reformulate the $square$-$root$ procedure as follows:

In [6]:
cat 1.3/Example_14/square_root_by_fixed_point.scm

(load "1.3/Example_13/fixed_point_of_function.scm")
(load "1.3/Example_14/average_damp.scm")
(define (square-root x)
  (fixed-point (average-damp (lambda (y) (/ x y))) 1.0))


### Running Instance:

Notice how this formulation makes explicit the three ideas in the method:fixed-point search, average damping, and the function $y \mapsto x/y$. It is instructive to compare this formulation of the $square$-$root$ method with the original version given in Section 1.1.7. Bear in mind that these procedures express the same process, and notice how much clearer the idea becomes when we express the process in terms of these abstractions. In
general, there are many ways to formulate a process as a procedure. Experienced programmers know how to choose procedural formulations that are particularly perspicuous, and where useful elements of the process are exposed as separate entities that can be reused in other applications. As a simple example of $reuse$, notice that the cube root of $x$ is a fixed point of the function $y \mapsto x/y^2$, so we can immediately generalize our square-root procedure to one that extracts cube roots:

In [7]:
cat 1.3/Example_14/cube_root_by_fixed_point.scm

(load "1.3/Example_13/fixed_point_of_function.scm")
(load "1.3/Example_14/average_damp.scm")
(define (cube-root x)
  (fixed-point (average-damp (lambda (y) (/ x (square y)))) 1.0))


## Example 15: Square Root --  Newton’s method

When we first introduced the $square$-$root$ procedure, in Section 1.1.7, we mentioned that this was a special case of $Newton’s\; method$.If $x\mapsto g(x)$ is a differentiable function, then a solution of the equation $g(x)\,=\,0$ is a fixed point of the function $x \mapsto f(x)$, where

$$f(x) = x - \frac{g(x)}{Dg(x)}$$

and $Dg(x)$ is the derivative of $g$ evaluated at $x$.

Newton’s method is the use of the fixed-point method we saw above to approximate a solution of the equation by finding a fixed point of the function $f$.

For many functions $g$ and for sufficiently good initial guesses for $x$, Newton’s method converges very rapidly to a solution of $g(x)\, =\, 0$.

In order to implement Newton’s method as a procedure, we must first express the idea of derivative. Note that “derivative,” like average damping, is something that transforms a function into another function. For instance, the derivative of the function $x \mapsto x^3$ is the function $x \mapsto 3x^2$. In general, if $g$ is a function and dx is a small number, then the derivative $Dg$ of $g$ is the function whose value at any number $x$ is given (in the limit of small $dx$) by

$$Dg(x) = \frac{g(x + dx) - g(x)}{dx}.$$

Thus, we can express the idea of $derivative$ (taking $dx$ to be, say, 0.00001) as the procedure

Like $average$-$damp$, $deriv$ is a procedure that takes a procedure as argument and returns a procedure as value. For example, to approximate the derivative of $x \mapsto x^3$ at 5 (whose exact value is 75) we can evaluate

With the aid of $deriv$, we can express Newton’s method as a fixed-point process:

The $newton$-$transform$ procedure expresses the formula at the beginning of this section, and newtons-method is readily defined in terms of this. It takes as arguments a procedure that computes the function for which we want to find a zero, together with an initial guess. For instance,to find the square root of x, we can use Newton’s method to find a zero of the function $y \mapsto y^2 - x$ starting with an initial guess of 1.

This provides yet another form of the $square$-$root$ procedure:

In [2]:
cat 1.3/Example_15/square_root_by_newton_method.scm

(load "1.3/Example_13/fixed_point_of_function.scm")

(define dx 0.00001)
(define (deriv g) (lambda (x) (/ (- (g (+ x dx)) (g x)) dx)))
(define (newton-transform g)
  (lambda (x) (- x (/ (g x) ((deriv g) x)))))
(define (newtons-method g guess)
  (fixed-point (newton-transform g) guess))
(define (square-root x)
  (newtons-method (lambda (y) (- (square y) x)) 1.0))


### Running Instance:

## Abstractions and first-class procedures

## Example 16: Square Root -- General Method

We’ve seen two ways to express the $square$-$root$ computation as an instance of a more general method, once as a $fixed$-$point\; search$ and once using $Newton’s\; method$. Since Newton’s method was itself expressed as a fixed-point process, we actually saw two ways to compute square roots as $fixed$ $points$. Each method begins with a function and finds a fixed point of some transformation of the function. We can express this general idea itself as a procedure:

In [11]:
cat 1.3/Example_16/fixed_point_of_transform.scm

(load "1.3/Example_13/fixed_point_of_function.scm")
(define (fixed-point-of-transform g transform guess)
  (fixed-point (transform g) guess))


This very general procedure takes as its arguments a procedure $g$ that computes some function, a procedure that transforms $g$, and an initial guess. The returned result is a fixed point of the transformed function.

Using this abstraction, we can recast the first $square$-$root$ computation from this section (where we look for a fixed point of the $average$-$damped$ version of $y \mapsto x/y$) as an instance of this general method:

In [13]:
cat 1.3/Example_16/square_root_by_fixed_point_general_method.scm

(load "1.3/Example_14/average_damp.scm")
(load "1.3/Example_16/fixed_point_of_transform.scm")
(define (square-root x)
  (fixed-point-of-transform
    (lambda (y) (/ x y)) average-damp 1.0))


### Running Instance:

Similarly, we can express the second $square$-$root$ computation from this section (an instance of $Newton’s\; method$ that finds a fixed point of the $Newton\; transform$ of $y \mapsto y^2-x$) as

In [15]:
cat 1.3/Example_16/square_root_by_newton_method_general_method.scm

(load "1.3/Example_16/newton_transform.scm")
(load "1.3/Example_16/fixed_point_of_transform.scm")
(define (square-root x)
  (fixed-point-of-transform
    (lambda (y) (- (square y) x)) newton-transform 1.0))


### Running Instance:

We began Section 1.3 with the observation that $compound$ $procedures$ are a crucial abstraction mechanism, because they permit us to express general methods of computing as explicit elements in our programming language. Now we’ve seen how $higher$-$order$ $procedures$ permit us to manipulate these general methods to create further abstractions.

As programmers, we should be alert to opportunities to identify the underlying abstractions in our programs and to build upon them and generalize them to create more powerful abstractions. This is not to say that one should always write programs in the most abstract way possible; expert programmers know how to choose the level of abstraction appropriate to their task. But it is important to be able to think in terms of these abstractions, so that we can be ready to apply them in new contexts. The significance of higher-order procedures is that they
enable us to represent these abstractions explicitly as elements in our programming language, so that they can be handled just like other computational elements.

In general, programming languages impose restrictions on the ways in which computational elements can be manipulated. Elements with the fewest restrictions are said to have $first$-$class$ status. Some of the “rights and privileges” of first-class elements are:

$\bullet$ They may be named by variables.

$\bullet$ They may be passed as arguments to procedures.

$\bullet$ They may be returned as the results of procedures.

$\bullet$ They may be included in data structures.

Lisp, unlike other common programming languages, awards procedures full first-class status. This poses challenges for efficient implementation,but the resulting gain in expressive power is enormous.

## Exercise 1.40: 
Define a procedure $\textbf{cubic}$ that can be used together with the $\textbf{newtons-method}$ procedure in expressions of the form

to approximate zeros of the cubic $x^3 + ax^2 + b^x + c$.

### Answer:

$\bullet$ First,we can define procedure $\textbf{cubic}$

$\bullet$ The $\textbf{newtons-method}$ procedure

In [3]:
cat 1.3/Exercise_1.40/newton_method.scm

(load "1.3/Example_13/fixed_point_of_function.scm")
(define dx 0.00001)
(define (deriv g) (lambda (x) (/ (- (g (+ x dx)) (g x)) dx)))
(define (newton-transform g)
  (lambda (x) (- x (/ (g x) ((deriv g) x)))))
(define (newtons-method g guess)
  (fixed-point (newton-transform g) guess))


$\bullet$ Then in terms of the $\textbf{newtons-method}$ procedure above,we can get $\textbf{zero-of-cubic}$ procedure

In [4]:
cat 1.3/Exercise_1.40/zero_of_cubic.scm

(load "1.3/Exercise_1.40/cubic.scm")
(load "1.3/Exercise_1.40/newton_method.scm")
(define (zero-of-cubic a b c)
  (newtons-method (cubic a b c) 1.0))


### Running Instance:

## Exercise 1.41: 
Define a procedure $\text{double}$ that takes a procedure of one argument as argument and returns a procedure that applies the original procedure twice. For example,if $\text{inc}$ is a procedure that adds 1 to its argument, then $\text{(double inc)}$ should be a procedure that adds 2. What value is returned by

### 解答：

对于一个接受单个参数$\textbf{x}$的函数$\textbf{f}$来说，要将它应用两次的方法是执行如下表达式

由此可以给出相应的$\textbf{double}$过程，它接受一个过程$\textbf{f}$，并且返回一个能将$\textbf{f}$应用两次的过程

In [6]:
cat 1.3/Exercise_1.41/double.scm

(define (double f) (lambda (x) (f (f x))))


### Running Instance:

## Exercise 1.42: 
Let ${f}$ and ${g}$ be two one-argument functions.The composition ${f}$ after ${g}$ is defined to be the function $x \mapsto f(g(x))$. Define a procedure $\text{compose}$ that implements composition.For example, if $inc$ is a procedure that adds 1 to its argument,

### 解答：

根据Example 1.41中$\text{double}$过程提示，我们能直接得到题目中要求的$\text{compose}$过程

In [7]:
cat 1.3/Exercise_1.42/compose.scm

(define (compose f g) (lambda (x) (f (g x))))


### Running Instance:

## Exercise 1.43: 
If $f$ is a numerical function and $n$ is a positive integer, then we can form the $n^{th}$ repeated application
of $f$, which is defined to be the function whose value at $x$ is $f(f(\dots(f(x))\dots))$. For example, if $f$ is the function $x \mapsto x + 1$,then the $n^{th}$ repeated application of $f$ is the function $x \mapsto x + n$. If $f$ is the operation of squaring a number,then the $n^{th}$ repeated application of $f$ is the function that raises its argument to the $\text{$2^n$-th}$ power. Write a procedure that takes as inputs a procedure that computes $f$ and a positive integer $n$ and returns the procedure that computes the $n^{th}$ repeated application of $f$. Your procedure should be able to be used as follows:

Hint: You may find it convenient to use $\text{compose}$ from Exercise 1.42.

## 解答：

对于表达式

$\text{repeated}$过程返回一个接受单个参数${x}$的过程g，在过程${g}$中，过程$f$被调用$n$次

比如表达式$\text{(repeated square 4)}$使用$\text{lambda}$展开为

根据上述递归计算过程，我们能定义出$\text{repeated}$递归过程

In [8]:
cat 1.3/Exercise_1.43/repeated_by_recursion.scm

(define (repeated f n)
  (if (= n 1) f
      (lambda (x) (f ((repeated f (- n 1)) x)))))


### Running Instance:

$\text{repeated}$过程的迭代版本

In [9]:
cat 1.3/Exercise_1.43/repeated_by_iteration.scm

(define (repeated f n)
  (define (iter i repeated-f)
    (if (= i 1) repeated-f
        (iter (- i 1) (lambda (x) (f (repeated-f x))))))
  (iter n f))


然而，根据题末给出的Hint，我们可以使用Example 1.42中的过程$\text{compose}$来定义我们的$\text{repeated}$过程

#### 递归版本

In [10]:
cat 1.3/Exercise_1.43/repeated_by_compose_recursion.scm

(load "1.3/Exercise_1.42/compose.scm")
(define (repeated f n)
  (if (= n 1) f
      (compose f (repeated f (- n 1)))))


#### 迭代版本

In [11]:
cat 1.3/Exercise_1.43/repeated_by_compose_iteration.scm

(load "1.3/Exercise_1.42/compose.scm")
(define (repeated f n)
  (define (iter i repeated-f)
    (if (= i 1) repeated-f
        (iter (- i 1) (compose f repeated-f))))
  (iter n f))


### Running Instance:

## Exercise 1.44:
The idea of $smoothing$ a function is an important concept in signal processing. If $f$ is a function and $dx$ is some small number, then the smoothed version of $f$ is the function whose value at a point $x$ is the average of $f(x-dx),f(x),f(x+dx)$.Write a procedure $smooth$ that takes as input a procedure that computes $f$ and returns a procedure that computes the smoothed $f$. It is sometimes valuable to repeatedly smooth a function (that is, smooth the smoothed function, and so on) to obtain the $n-fold\;smoothed function$. Show how to generate the n-fold smoothed function of any given function using $smooth$ and $repeated$ from Exercise 1.43.

## 解答：

$\clubsuit$ 根据题中给出的关于平滑函数的说明，我们能直接定义$smooth$过程

In [12]:
cat 1.3/Exercise_1.44/smooth.scm

(define dx 0.00001)
(define (smooth f)
  (lambda (x) (/ (+ (f (- x dx)) (f x) (f (+ x dx))) 3)))


### Running Instance:

$\clubsuit$ 对于连续n次平滑函数，根据如下表达式的展开

直观地得到$\text{smooth-n-times}$过程的递归版本定义

In [13]:
cat 1.3/Exercise_1.44/smooth_n_times_by_recursion.scm

(load "1.3/Exercise_1.44/smooth.scm")
(define (smooth-n-times f n)
  (if (= n 0) f (smooth (smooth-n-times f (- n 1)))))


而$\text{smooth-n-times}$过程的迭代版本定义

In [14]:
cat 1.3/Exercise_1.44/smooth_n_times_by_iteration.scm

(load "1.3/Exercise_1.44/smooth.scm")
(define (smooth-n-times f n)
  (define (iter i smoothed-f)
    (if (= i 0) smoothed-f
        (iter (- i 1) (smooth smoothed-f))))
  (iter n f))


### Running Instance:

$\clubsuit$ 使用$\text{repeated}$定义$\text{smooth-n-times}$

对于表达式$\text{(repeated smooth 3)}$的$lambda$展开

有上述计算过程，我们能得到使用$\text{repeated}$定义$\text{smooth-n-times}$的版本  
(依据使用的repeated过程的版本来确定递归还是迭代)

In [15]:
cat 1.3/Exercise_1.44/smooth_n_times_by_repeated_recursion.scm

(load "1.3/Exercise_1.43/repeated_by_recursion.scm")
(load "1.3/Exercise_1.44/smooth.scm")
(define (smooth-n-times f n)
  (let ((n-times-smooth (repeated smooth n)))
    (n-times-smooth f)))


### Running Instance:

## Exercise 1.45: 
We saw in Section 1.3.3 that attempting to compute square roots by naively finding a fixed point of $y \mapsto x/y$ does not converge, and that this can be fixed by average damping. The same method works for finding cube
roots as fixed points of the average-damped $y \mapsto x/y^2$. Unfortunately,the process does not work for fourth roots—a single average damp is not enough to make a fixed-point search for $y \mapsto x/y^3$ converge. On the other hand, if we average damp twice (i.e., use the average damp of the average damp of $y \mapsto x/y^3$) the fixed-point search does converge. Do some experiments to determine how many average damps are required to compute $n^{th}$ roots as a fixedpoint search based upon repeated average damping of $y \mapsto x/y^{n-1}$. Use this to implement a simple procedure for computing $n^{th}$ roots using $\text{fixed-point}$, $\text{average-damp}$, and the $\text{repeated}$ procedure of Exercise 1.43. Assume that any arithmetic operations you need are available as primitives.

$\spadesuit$ 使用$\text{repeated}$过程重新定义$\text{expt}$过程

$\text{expt}$用于计算$y^{n-1}$，接受两个参数$base$和$n$，计算出$base^n$

我们调用$\text{repeated}$过程来计算表达式$2^5$的替换模型:

根据上述计算过程，我们能够直接定义$\text{expt}$过程

In [16]:
cat 1.3/Exercise_1.45/expt.scm

(load "1.3/Exercise_1.43/repeated_by_recursion")
(define (expt base n)
  (if (= n 0) 1 ((repeated (lambda (x) (* base x)) n) 1)))


$\spadesuit$ 使用$\text{repeated}$过程定义可进行不定次数平均阻尼的过程$\text{average-damp-n-times}$

因为需要对输入的公式进行不定次数的$\text{average-damp}$变换以确保不动点收敛，所以我们需要定义$\text{average-damp-n-times}$  
$\text{average-damp-n-times}$过程接受参数$f$和$n$，返回一个经过$n$次$\text{average-damp}$变换的$f$作为返回值

In [22]:
cat 1.3/Exercise_1.45/average_damp_n_times.scm

(load "1.3/Exercise_1.43/repeated_by_recursion")
(load "1.3/Example_14/average_damp.scm")
(define (average-damp-n-times f n)
  ((repeated average-damp n) f))


$\spadesuit$ 使用上述的$\text{expt}$过程，$\text{average-damp-n-times}$过程以及$\text{fixed-point}$过程，我们能够定义求$n$次方根的过程$\text{dampd-nth-root}$ 

过程$\text{dampd-nth-root}$接受两个参数$n$和$damp$-$times$：  
$n$表示要计算的方根次数，  
$damp$-$times$指定要对公式进行多少次平均阻尼变换  
返回值是一个过程，该过程接受参数$x$，返回$x$的$n$次方根

In [23]:
cat 1.3/Exercise_1.45/damped_nth_root.scm

(load "1.3/Example_13/fixed_point_of_function.scm")
(load "1.3/Exercise_1.45/expt.scm")
(load "1.3/Exercise_1.45/average_damp_n_times.scm")
(define (damped-nth-root n damp-times)
  (lambda (x) (fixed-point (average-damp-n-times (lambda (y) (/ x (expt y (- n 1)))) damp-times) 1.0)))


通过定义平方根、立方根和四次方根来测试$\text{damped-nth-root}$过程

### Running Instance:

$\bullet$ 当只是调用一次平均阻尼变换时，计算四次方根时不动点序列已经不收敛

因此我们增加一次平均阻尼变换

$\spadesuit$ 确定关于计算n次方根和不动点收敛所需的平均阻尼变换次数之间的关系

通过上述过程我们可以得到如下一些测试数据

可以看出，要使得计算$n$次方根的不动点收敛，最少需要$lgn$次平均阻尼变换

所以我们需要定义$lg$过程

In [24]:
cat 1.3/Exercise_1.45/lg.scm

(define (lg n)
  (cond ((> (/ n 2) 1) (+ 1 (lg (/ n 2))))
        ((< (/ n 2) 1) 0)
        (else 1)))


$\spadesuit$ 定义$\text{nth-root}$过程

In [25]:
cat 1.3/Exercise_1.45/nth-root.scm

(load "1.3/Exercise_1.45/damped_nth_root.scm")
(load "1.3/Exercise_1.45/lg.scm")
(define (nth-root n) (damped-nth-root n (lg n)))


### Running Instance:

## Exercise 1.46: 
Several of the numerical methods described in this chapter are instances of an extremely general computational
strategy known as $\textbf{iterative improvement}$.Iterative improvement says that, to compute something, we start with an initial guess for the answer, test if the guess is good enough, and otherwise improve the guess and continue the process using the improved guess as the new guess. Write a procedure $\text{iterative-improve}$ that takes two procedures as arguments: a method for telling whether a guess is good enough and a method for improving a guess. $\text{iterative-improve}$ should return as its value a procedure that takes a guess as argument and keeps improving the guess until it is good enough. Rewrite the $\text{square-root}$ procedure of Section 1.1.7 and the $\text{fixed-point}$ procedure of Section 1.3.3 in terms of $\text{iterative-improve}$.

根据描述，我们给出如下形式的过程

而Example_13中$\text{fixed-point}$过程

In [26]:
cat 1.3/Example_13/fixed_point_of_function.scm

(define tolerance 0.00001)
(define (fixed-point f first-guess)
  (define (close-enough v1 v2)
    (< (abs (- v1 v2)) tolerance))
  (define (try guess)
    (let ((next (f guess)))
      (if (close-enough guess next)
          next
          (try next))))
  (try first-guess))


$\text{fixed-point}$过程不仅仅和$\text{iterative-improve}$非常相似，事实上，$\text{iterative-improve}$就隐藏在$\text{fixed-point}$当中

在$\text{fixed-point}$中，$\text{close-enough?}$负责检查猜测是否足够好，而函数$f$负责改进猜测，如果我们将$\text{close-enough?}$过程抽取出来，成为额外的参数，那么$\text{fixed-point}$过程的定义就变成了

如果再将$\text{first-guess}$从$\text{fixed-point}$过程的参数列表中移走，变成另一个包裹$\text{try}$过程的$lambda$ 的参数，$\text{fixed-point}$过程的定义

现在的$\text{first-guess}$定义已经和前面给出的$\text{iterative-improve}$形式一样了，如果将$\text{fixed-point}$改名成$\text{iterative-improve}$，将参数$f$改名成$improve$，并且删除$\text{close-enough?}$过程和$dx$变量，题目要求的$\text{iterative-improve}$就出现了

In [27]:
cat 1.3/Exercise_1.46/iterative_improve.scm

(define (iterative-improve close-enough? improve)
  (lambda (first-guess)
    (define (try guess)
      (let ((next (improve guess)))
           (if (close-enough? guess next)
               next
               (try next)))) 
    (try first-guess)))


$\spadesuit$  用$\text{iterative-improve}$重新定义$\text{fixed-point}$

In [28]:
cat 1.3/Exercise_1.46/fixed_point_by_iterative_improve.scm

(load "1.3/Exercise_1.46/iterative_improve.scm")

(define (fixed-point f first-guess)
  (define tolerance 0.00001)
  (define (close-enough? v1 v2)
    (< (abs (- v1 v2)) tolerance))
  (define (improve guess)
    (f guess))
  ((iterative-improve close-enough? improve) first-guess))


### Running Instance:

$\spadesuit$  用$\text{iterative-improve}$重新定义$\text{square-root}$

In [29]:
cat 1.3/Exercise_1.46/square_root_by_iterative_improve.scm

(load "1.3/Exercise_1.46/iterative_improve.scm")

(define (square-root x)
  (define dx 0.00001)
  (define (close-enough? v1 v2)
    (< (abs (- v1 v2)) dx))
  (define (improve guess)
    (average guess (/ x guess)))
  (define (average x y)
    (/ (+ x y) 2))
  ((iterative-improve close-enough? improve) 1.0))


### Running Instance: